src/python/pants/backend/python/dependency_inference/rules.py

```diff
@@ -95,23 +110,28 @@ async def infer_python_dependencies(
         for fp in stripped_sources.snapshot.files
     )
     digest_contents = await Get(DigestContents, Digest, stripped_sources.snapshot.digest)
-    imports_per_file = tuple(
-        find_python_imports(file_content.content.decode(), module_name=module.module)
-        for file_content, module in zip(digest_contents, modules)
-    )
-    owner_per_import = await MultiGet(
-        Get(PythonModuleOwner, PythonModule(imported_module))
-        for file_imports in imports_per_file
-        for imported_module in file_imports.explicit_imports
-        if imported_module not in combined_stdlib
-    )
+
+    owner_requests: List[Get[PythonModuleOwner, PythonModule]] = []
+    for file_content, module in zip(digest_contents, modules):
+        file_imports_obj = find_python_imports(
+            file_content.content.decode(), module_name=module.module
+        )
+        detected_imports = (
+            file_imports_obj.all_imports
+            if python_inference.string_imports
+            else file_imports_obj.explicit_imports
+        )
+        owner_requests.extend(
+            Get(PythonModuleOwner, PythonModule(imported_module))
+            for imported_module in detected_imports
+            if imported_module not in combined_stdlib
+        )
+
+    owner_per_import = await MultiGet(owner_requests)
     result = (
         owner.address
         for owner in owner_per_import
-        if (
-            owner.address
-            and owner.address.maybe_convert_to_base_target() != request.sources_field.address
```

Comment:
Originally, we would not infer for a base target any dependencies on its own generated subtargets. That used to make sense. Now, though, with the model change, it makes sense for a base target to have a dependency on its generated subtarget.